bin/fetch-configlet

```diff
@@ -23,21 +33,40 @@ def get_os():
 
 
 def get_arch():
-    return '64bit' if sys.maxsize > 2**32 else '32bit'
+    return "64bit" if sys.maxsize > 2 ** 32 else "32bit"
 
 
 def fetch_configlet():
-    latest = "https://api.github.com/repos/exercism/configlet/releases/latest"
-    resp = requests.get(latest)
-    data = resp.json()
+    with urlopen(
+        "https://api.github.com/repos/exercism/configlet/releases/latest"
+    ) as request:
+        status = request.status
+        headers = dict(request.getheaders())
+        response = request.read()
+    if status >= 500:
+        print(f"Sleeping due to {status} response from API.")
+        time.sleep(5)
+        return 1
+    if status == 403:
+        wait_until = datetime.fromtimestamp(
+            headers["X-RateLimit-Reset"], timezone.utc
+        )
+        delta = wait_until - datetime.now(timezone.utc)
+        seconds = delta.total_seconds()
+        wait = seconds + 5 if seconds > 5 else 5
+        print(f"Rate limited, sleeping {wait} seconds.")
+        time.sleep(wait)
+        return 1
+    if status >= 400:
+        sys.exit(f"Received unexpected {status} response from API.")
+    data = json.loads(response.decode())
     version = data["tag_name"]
     machine_info = f"{get_os()}-{get_arch()}"
     name = f"configlet-{machine_info}.tgz"
     for asset in data["assets"]:
         if asset["name"] == name:
             print(f"Downloading configlet {version} for {machine_info}")
-            download_and_extract(asset["browser_download_url"])
-            return 0
+            return download_and_extract(asset["browser_download_url"])
```

Comment:
For the sake of argument let's assume there's a temporary hiccup server side that will cause this to return 1 on the first attempt to download one of the assets ... this hiccup _will_ be fixed by a 5 second wait; _should_ we repeat the effort of fetching **data** and finding the appropriate asset again?

Basically shouldn't **download_and_extract** have its own backoff loop?

Reply:
My thought was that we can't guarantee what the API will do in the future and race conditions. They might switch to temporary URLs, or the latest version change out from under us, etc.

Reply:
I should think that if they’ve made an API change such that between parsing the configlet version and finding the platform-specific link for that version they’ve changed the endpoint of the URLs they’ve just served us then we’ve got bigger problems that aren’t going to be fixed by parsing the version number again.

I can see the race condition point, but if the latest version was swapped out form under us via a normal process it would leave the old download URLs intact, so your script would deliver stale data anyway... if the link has been permanently invalidated then retrying the download on its own loop will eventually kick back to
the outer loop and pick up the new version anyways, it just won’t needlessly hammer the latest endpoint — the one we already have rate limit issues with — in the more likely case, when there’s some temporary issue that’s blocking a valid URL from working.

Reply:
I'll just change it to use MAX_RETRIES in the method then.